src/main/java/com/enderio/base/common/handler/GrindingHandler.java

```diff
@@ -3,53 +3,57 @@
+        if (!(player.isCrouching())
+            || (target != Blocks.OBSIDIAN && target != Blocks.GRINDSTONE && target != Blocks.CRYING_OBSIDIAN)
+            || !(offhand.is(Items.FLINT))) {
+            return;
         }
-    }
-
-    @SubscribeEvent
-    public static void onGrindstoneResult(GrindstoneEvent.OnTakeItem event) {
-        if ((event.getTopItem().is(Items.DEEPSLATE) || event.getTopItem().is(Items.COBBLED_DEEPSLATE)) && event.getBottomItem().is(Items.FLINT)) {
-            ItemStack top = event.getTopItem().copy();
-            ItemStack bottom = event.getBottomItem().copy();
-
-            top.shrink(1);
-
-            // 33% chance to destroy the flint
-            if (RandomSource.create().nextInt(3) == 0) {
-                bottom.shrink(1);
+        if (mainhand.is(Items.DEEPSLATE) || mainhand.is(Items.COBBLED_DEEPSLATE)) {
+            mainhand.shrink(1);
+            // Grindstone consumes Flint 33% of the time
+            if (target != Blocks.OBSIDIAN && target != Blocks.CRYING_OBSIDIAN && RandomSource.create().nextInt(3) == 0) {
+                offhand.shrink(1);
             }
-
-            event.setNewTopItem(top);
-            event.setNewBottomItem(bottom);
-        } else if (event.getTopItem().is(Items.COAL) && event.getTopItem().getCount() >= 3 && event.getBottomItem().is(Items.FLINT)) {
-            ItemStack top = event.getTopItem().copy();
-            ItemStack bottom = event.getBottomItem().copy();
-
-            top.shrink(3);
-
-            // 33% chance to destroy the flint
-            if (RandomSource.create().nextInt(3) == 0) {
-                bottom.shrink(1);
+            int output = 1;
+            // Crying Obsidian has a 33% chance to double the output
+            if (target == Blocks.CRYING_OBSIDIAN && RandomSource.create().nextInt(3) == 0) {
+                output = 2;
             }
-
-            event.setNewTopItem(top);
-            event.setNewBottomItem(bottom);
+            player.addItem(new ItemStack(EIOItems.GRAINS_OF_INFINITY.get(), output));
+            event.setCanceled(true);
+        } else if (mainhand.is(Items.COAL) && mainhand.getCount() >= 3) {
+            int cost = 3;
+            // Crying Obsidian had a 33% chance to only consume 1 coal instead of 3
+            if (target == Blocks.CRYING_OBSIDIAN && RandomSource.create().nextInt(3) == 0) {
+                cost = 1;
+            }
+            mainhand.shrink(cost);
+            // Grindstone consumes Flint 33% of the time
+            if (target != Blocks.OBSIDIAN && target != Blocks.CRYING_OBSIDIAN && RandomSource.create().nextInt(3) == 0) {
+                offhand.shrink(1);
+            }
+            player.addItem(new ItemStack(EIOItems.POWDERED_COAL.get(), 1));
+        }
+        // Cancel the event so that we don't bring up the GUI for the grindstone. Unnecessary for obsidian
+        if (target == Blocks.GRINDSTONE) {
+            event.setCanceled(true);
         }
```

Comment:
cancel it for all cases, to mark that we have done something and others don't do further interactions